mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -882,5 +900,79 @@ LogicalResult tuningTableLookup(TuningTable *perfTable, ModuleOp &mod,
   return failure();
 }
 
+template <typename ParamType>
+int64_t retrieveSplitKValueImpl(StringRef perfConfig) {
+  ParamType params;
+  params.deserialize(perfConfig.str());
+  return params.splitKFactor;
+}
+
+int64_t retrieveSplitKValue(rock::RockGemmWrapperInterface op,
+                            StringRef perfConfig) {
+  rock::GemmFeatures features = op.getGemmFeatures();
+  if (isAccel(features)) {
+    return retrieveSplitKValueImpl<rock::InitParamsAccel>(perfConfig);
+  }
+  return retrieveSplitKValueImpl<rock::InitParamsNonAccel>(perfConfig);
+}
+
+bool isSplitKRequested(ModuleOp &mod, StringRef &perfConfig) {
+  WalkResult gemmWalkResult =
+      mod.walk([&](rock::RockGemmWrapperInterface op) -> WalkResult {
+        int64_t splitKFactor = retrieveSplitKValue(op, perfConfig);
+        if (splitKFactor > 1) {
+          return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+      });
+
+  return gemmWalkResult.wasInterrupted();
+}
+
+RocmlirSplitKSelectionLikelihood isSplitKFaster(int64_t gDim, int64_t mDim,
+                                                int64_t nDim, int64_t kDim,
+                                                int64_t numCUs) {
+
+  // M/block N/block K/block M/wave N/wave kPack
+  const std::vector<std::vector<uint32_t>> rangeGemmParams = {
+      {4, 8, 16, 32, 64, 128, 256},
+      {16, 32, 64, 128, 256},
+      {1, 2, 4, 8},
+      {1, 4, 8, 16}};
+
+  rock::GemmSize gemmSize(gDim, mDim, kDim, nDim);
+  llvm::SmallSetVector<int64_t, 10> splitKValues = {};
+  double minWorkImbalance = std::numeric_limits<double>::max();
+  for (uint32_t mPerBlock : rangeGemmParams[0]) {
+    for (uint32_t nPerBlock : rangeGemmParams[1]) {
+      for (uint32_t kPerBlock : rangeGemmParams[2]) {
+        for (uint32_t kPack : rangeGemmParams[3]) {
+          const double currWorkImbalance = computeWorkImbalance(
+              gemmSize, mPerBlock, nPerBlock, kPerBlock, kPack, numCUs);
+          minWorkImbalance = std::min(currWorkImbalance, minWorkImbalance);
+
+          llvm::SmallVector<int64_t> currSplitKValues =
+              computeOptimalSplitKFactors(gemmSize, mPerBlock, nPerBlock,
+                                          kPerBlock, kPack, numCUs);
+          std::for_each(
+              currSplitKValues.begin(), currSplitKValues.end(),
+              [&splitKValues](int64_t value) { splitKValues.insert(value); });
+        }
+      }
+    }
+  }
+
+  if (splitKValues.size() == 1) {
```

Comment:
This'll never hit, I think, since each calculation returns, at minimum, a vector containing `1`.

Perhaps the `SmallVector` above should be some sort of binary heap situation?

Reply:
Wait, `splitKValues` has type `llvm::SmallSetVector`. So, it should work